File: ascii.py
import time
import cv2
import numpy as np
import os
ascii_chars = tuple(reversed('$@B%8&WM#*oahkbdpqwmZO0QLCJUYXzcvunxrjft/|()1{}[]?-_+~<>i!lI;:,"^`\'.    '))

file_path = input("Enter the file path: ")
if file_path == "0":
    file_path = 0
fps = int(input("Enter FPS: "))

vc = cv2.VideoCapture(file_path)
while True:
    rval, frame = vc.read()
    resized_frame = cv2.resize(frame, (tuple(os.get_terminal_size())[0]-1, tuple(os.get_terminal_size())[1]))
    img_array = np.asarray(cv2.flip(cv2.cvtColor(resized_frame, cv2.COLOR_BGR2GRAY), 1))
    print('\n'.join(''.join(ascii_chars[round((len(ascii_chars) - 1) * color / 255)] for color in row) for row in img_array), end="\r")
    time.sleep(1/fps)
vc.release()

File: ascii.cpp
#include <opencv2/opencv.hpp>
#include <iostream>
#include <windows.h>
#include <cstdlib>

using namespace std;
using namespace cv;

int main() {
    Mat frame;
    char ascii_chars[] = {' ', ' ', ' ', ' ', '.', '\'', '`', '^', '"', ',', ':', ';', 'I', 'l', '!', 'i', '>', '<', '~', '+', '_', '-', '?', ']', '[', '}', '{', '1', ')', '(', '|', '/', 't', 'f', 'j', 'r', 'x', 'n', 'u', 'v', 'c', 'z', 'X', 'Y', 'U', 'J', 'C', 'L', 'Q', '0', 'O', 'Z', 'm', 'w', 'q', 'p', 'd', 'b', 'k', 'h', 'a', 'o', '*', '#', 'M', 'W', '&', '8', '%', 'B', '@', '$'};
    string file_path;
    string fps;
    cout << "Enter a path for the video (0 for webcam): ";
    cin >> file_path;
    cout << "Enter FPS: ";
    cin >> fps;
    float fps_delay = 1 / stof(fps) * 1000;
    VideoCapture cap;
    if (file_path == "0") {
        cap.open(0);
    } else {
        cap.open(file_path);
    }
    if (!cap.isOpened()) {
        cout << "Error opening video stream or file" << endl;
        return -1;
    }

    CONSOLE_SCREEN_BUFFER_INFO csbi;
    int columns, rows;

    while (true){
        GetConsoleScreenBufferInfo(GetStdHandle(STD_OUTPUT_HANDLE), &csbi);
        columns = csbi.srWindow.Right - csbi.srWindow.Left;
        rows = csbi.srWindow.Bottom - csbi.srWindow.Top;

        cap >> frame;
        if (frame.empty()) {
            break;
        }

        resize(frame, frame, Size (columns, rows));
        cvtColor(frame, frame, COLOR_BGR2GRAY);
        Size s = frame.size();
        string frame_text = "";

        Sleep(fps_delay);
        cout << fps_delay << " " << fps << endl;

        for (int y=0; y < s.height; y++) {
            for (int x=0; x < s.width; x++) {
                frame_text += ascii_chars[(sizeof(ascii_chars) - 1) * frame.at<uchar>(y, x) / 255];
            }
            frame_text += "\n";
        }
        cout << frame_text;
    }
    cap.release();
    return 0;
}
